// @apollo/react-testing/lib/react-testing.umd.js
(function (global, factory) {
  if (typeof define === "function" && define.amd) {
    define(["exports", "tslib", "react", "apollo-client", "apollo-cache-inmemory", "@apollo/react-common", "apollo-link", "apollo-utilities", "graphql/language/printer", "fast-json-stable-stringify"], factory);
  } else if (typeof exports !== "undefined") {
    factory(exports, require("tslib"), require("react"), require("apollo-client"), require("apollo-cache-inmemory"), require("@apollo/react-common"), require("apollo-link"), require("apollo-utilities"), require("graphql/language/printer"), require("fast-json-stable-stringify"));
  } else {
    var mod = {
      exports: {}
    };
    factory(mod.exports, global.tslib, global.react, global.apolloClient, global.apolloCacheInmemory, global.reactCommon, global.apolloLink, global.apolloUtilities, global.printer, global.fastJsonStableStringify);
    global.unknown = mod.exports;
  }
})(this, function (_exports, _tslib, _react, _apolloClient, _apolloCacheInmemory, _reactCommon, _apolloLink, _apolloUtilities, _printer, _fastJsonStableStringify) {

  _exports.__esModule = true;
  _exports.createClient = createClient;
  _exports.mockObservableLink = mockObservableLink;
  _exports.mockSingleLink = mockSingleLink;
  _exports.stripSymbols = stripSymbols;
  _exports.wait = wait;
  _exports.MockedProvider = _exports.MockSubscriptionLink = _exports.MockLink = void 0;
  _react = _interopRequireDefault(_react);
  _fastJsonStableStringify = _interopRequireDefault(_fastJsonStableStringify);

  function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

  function requestToKey(request, addTypename) {
    var queryString = request.query && (0, _printer.print)(addTypename ? (0, _apolloUtilities.addTypenameToDocument)(request.query) : request.query);
    var requestKey = {
      query: queryString
    };
    return JSON.stringify(requestKey);
  }

  var MockLink = function (_super) {
    (0, _tslib.__extends)(MockLink, _super);

    function MockLink(mockedResponses, addTypename) {
      if (addTypename === void 0) {
        addTypename = true;
      }

      var _this = _super.call(this) || this;

      _this.addTypename = true;
      _this.mockedResponsesByKey = {};
      _this.addTypename = addTypename;
      if (mockedResponses) mockedResponses.forEach(function (mockedResponse) {
        _this.addMockedResponse(mockedResponse);
      });
      return _this;
    }

    MockLink.prototype.addMockedResponse = function (mockedResponse) {
      var normalizedMockedResponse = this.normalizeMockedResponse(mockedResponse);
      var key = requestToKey(normalizedMockedResponse.request, this.addTypename);
      var mockedResponses = this.mockedResponsesByKey[key];

      if (!mockedResponses) {
        mockedResponses = [];
        this.mockedResponsesByKey[key] = mockedResponses;
      }

      mockedResponses.push(normalizedMockedResponse);
    };

    MockLink.prototype.request = function (operation) {
      var key = requestToKey(operation, this.addTypename);
      var responseIndex;
      var response = (this.mockedResponsesByKey[key] || []).find(function (res, index) {
        var requestVariables = operation.variables || {};
        var mockedResponseVariables = res.request.variables || {};

        if (!(0, _apolloUtilities.isEqual)((0, _fastJsonStableStringify.default)(requestVariables), (0, _fastJsonStableStringify.default)(mockedResponseVariables))) {
          return false;
        }

        responseIndex = index;
        return true;
      });

      if (!response || typeof responseIndex === 'undefined') {
        throw new Error("No more mocked responses for the query: " + (0, _printer.print)(operation.query) + ", variables: " + JSON.stringify(operation.variables));
      }

      this.mockedResponsesByKey[key].splice(responseIndex, 1);
      var result = response.result,
          error = response.error,
          delay = response.delay,
          newData = response.newData;

      if (newData) {
        response.result = newData();
        this.mockedResponsesByKey[key].push(response);
      }

      if (!result && !error) {
        throw new Error("Mocked response should contain either result or error: " + key);
      }

      return new _apolloLink.Observable(function (observer) {
        var timer = setTimeout(function () {
          if (error) {
            observer.error(error);
          } else {
            if (result) {
              observer.next(typeof result === 'function' ? result() : result);
            }

            observer.complete();
          }
        }, delay ? delay : 0);
        return function () {
          clearTimeout(timer);
        };
      });
    };

    MockLink.prototype.normalizeMockedResponse = function (mockedResponse) {
      var newMockedResponse = (0, _apolloUtilities.cloneDeep)(mockedResponse);
      newMockedResponse.request.query = (0, _apolloUtilities.removeConnectionDirectiveFromDocument)(newMockedResponse.request.query);
      var query = (0, _apolloUtilities.removeClientSetsFromDocument)(newMockedResponse.request.query);

      if (query) {
        newMockedResponse.request.query = query;
      }

      return newMockedResponse;
    };

    return MockLink;
  }(_apolloLink.ApolloLink);

  _exports.MockLink = MockLink;

  function mockSingleLink() {
    var mockedResponses = [];

    for (var _i = 0; _i < arguments.length; _i++) {
      mockedResponses[_i] = arguments[_i];
    }

    var maybeTypename = mockedResponses[mockedResponses.length - 1];
    var mocks = mockedResponses.slice(0, mockedResponses.length - 1);

    if (typeof maybeTypename !== 'boolean') {
      mocks = mockedResponses;
      maybeTypename = true;
    }

    return new MockLink(mocks, maybeTypename);
  }

  var MockedProvider = function (_super) {
    (0, _tslib.__extends)(MockedProvider, _super);

    function MockedProvider(props) {
      var _this = _super.call(this, props) || this;

      var _a = _this.props,
          mocks = _a.mocks,
          addTypename = _a.addTypename,
          defaultOptions = _a.defaultOptions,
          cache = _a.cache,
          resolvers = _a.resolvers,
          link = _a.link;
      var client = new _apolloClient.ApolloClient({
        cache: cache || new _apolloCacheInmemory.InMemoryCache({
          addTypename: addTypename
        }),
        defaultOptions: defaultOptions,
        link: link || new MockLink(mocks || [], addTypename),
        resolvers: resolvers
      });
      _this.state = {
        client: client
      };
      return _this;
    }

    MockedProvider.prototype.render = function () {
      var _a = this.props,
          children = _a.children,
          childProps = _a.childProps;
      return children ? _react.default.createElement(_reactCommon.ApolloProvider, {
        client: this.state.client
      }, _react.default.cloneElement(_react.default.Children.only(children), (0, _tslib.__assign)({}, childProps))) : null;
    };

    MockedProvider.prototype.componentWillUnmount = function () {
      this.state.client.stop();
    };

    MockedProvider.defaultProps = {
      addTypename: true
    };
    return MockedProvider;
  }(_react.default.Component);

  _exports.MockedProvider = MockedProvider;

  var MockSubscriptionLink = function (_super) {
    (0, _tslib.__extends)(MockSubscriptionLink, _super);

    function MockSubscriptionLink() {
      var _this = _super.call(this) || this;

      _this.unsubscribers = [];
      _this.setups = [];
      return _this;
    }

    MockSubscriptionLink.prototype.request = function (_req) {
      var _this = this;

      return new _apolloLink.Observable(function (observer) {
        _this.setups.forEach(function (x) {
          return x();
        });

        _this.observer = observer;
        return function () {
          _this.unsubscribers.forEach(function (x) {
            return x();
          });
        };
      });
    };

    MockSubscriptionLink.prototype.simulateResult = function (result, complete) {
      var _this = this;

      if (complete === void 0) {
        complete = false;
      }

      setTimeout(function () {
        var observer = _this.observer;
        if (!observer) throw new Error('subscription torn down');
        if (complete && observer.complete) observer.complete();
        if (result.result && observer.next) observer.next(result.result);
        if (result.error && observer.error) observer.error(result.error);
      }, result.delay || 0);
    };

    MockSubscriptionLink.prototype.onSetup = function (listener) {
      this.setups = this.setups.concat([listener]);
    };

    MockSubscriptionLink.prototype.onUnsubscribe = function (listener) {
      this.unsubscribers = this.unsubscribers.concat([listener]);
    };

    return MockSubscriptionLink;
  }(_apolloLink.ApolloLink);

  _exports.MockSubscriptionLink = MockSubscriptionLink;

  function mockObservableLink() {
    return new MockSubscriptionLink();
  }

  function createClient(data, query, variables) {
    if (variables === void 0) {
      variables = {};
    }

    return new _apolloClient.ApolloClient({
      link: mockSingleLink({
        request: {
          query: query,
          variables: variables
        },
        result: {
          data: data
        }
      }),
      cache: new _apolloCacheInmemory.InMemoryCache({
        addTypename: false
      })
    });
  }

  function stripSymbols(data) {
    return JSON.parse(JSON.stringify(data));
  }

  function wait(ms) {
    return new Promise(function (resolve) {
      return setTimeout(resolve, ms);
    });
  } 

});
